Detectors/Calibration/SpacePoints/src/TrackResiduals.cxx

```diff
@@ -0,0 +1,1411 @@
+          cmatD[iMat++] += wi * dxi2 * dxi2;
+          rhsD[iRhs++] += wi * dxi2 * vi;
+        }
+        if (mSmoothPol2[param::VoxF]) {
+          cmatD[iMat++] += wi * dfi2;
+          cmatD[iMat++] += wi * dxi * dfi2;
+          cmatD[iMat++] += wi * dfi * dfi2;
+          cmatD[iMat++] += wi * dzi * dfi2;
+          cmatD[iMat++] += wi * dxi2 * dfi2;
+          cmatD[iMat++] += wi * dfi2 * dfi2;
+          rhsD[iRhs++] += wi * dfi2 * vi;
+        }
+        if (mSmoothPol2[param::VoxZ]) {
+          cmatD[iMat++] += wi * dzi2;
+          cmatD[iMat++] += wi * dxi * dzi2;
+          cmatD[iMat++] += wi * dfi * dzi2;
+          cmatD[iMat++] += wi * dzi * dzi2;
+          cmatD[iMat++] += wi * dxi2 * dzi2;
+          cmatD[iMat++] += wi * dfi2 * dzi2;
+          cmatD[iMat++] += wi * dzi2 * dzi2;
+          rhsD[iRhs++] += wi * dzi2 * vi;
+        }
+      }
+    }
+
+    bool fitRes = true;
+
+    // solve system of linear equations
+
+    TMatrixDSym matrix(matSize);
+    TDecompChol chol(matSize);
+    TVectorD rhsVec(matSize);
+    for (int iDim = 0; iDim < param::ResDim; ++iDim) {
+      if (!doDim[iDim]) {
+        continue;
+      }
+      matrix.Zero(); // reset matrix
+      double* cmatD = cmat[iDim];
+      double* rhsD = &mLastSmoothingRes[iDim * param::MaxSmtDim];
+      short iMat = -1;
+      short iRhs = -1;
+      short row = -1;
+
+      // with the studid implementation of TMatrixDSym we need to set all elements of the matrix explicitly (or maybe only upper triangle?)
+      matrix(++row, 0) = cmatD[++iMat];
+      matrix(++row, 0) = cmatD[++iMat];
+      matrix(row, 1) = cmatD[++iMat];
+      matrix(0, row) = matrix(row, 0);
+      matrix(++row, 0) = cmatD[++iMat];
+      matrix(row, 1) = cmatD[++iMat];
+      matrix(row, 2) = cmatD[++iMat];
+      matrix(0, row) = matrix(row, 0);
+      matrix(1, row) = matrix(row, 1);
+      matrix(++row, 0) = cmatD[++iMat];
+      matrix(row, 1) = cmatD[++iMat];
+      matrix(row, 2) = cmatD[++iMat];
+      matrix(row, 3) = cmatD[++iMat];
+      matrix(0, row) = matrix(row, 0);
+      matrix(1, row) = matrix(row, 1);
+      matrix(2, row) = matrix(row, 2);
+      // add pol2 elements if needed
+      if (mSmoothPol2[param::VoxX]) {
+        const unsigned int colLim = (++row) + 1;
+        for (int iCol = 0; iCol < colLim; ++iCol) {
+          matrix(row, iCol) = cmatD[++iMat];
+          matrix(iCol, row) = matrix(row, iCol);
+        }
+      }
+      if (mSmoothPol2[param::VoxF]) {
+        const unsigned int colLim = (++row) + 1;
+        for (int iCol = 0; iCol < colLim; ++iCol) {
+          matrix(row, iCol) = cmatD[++iMat];
+          matrix(iCol, row) = matrix(row, iCol);
+        }
+      }
+      if (mSmoothPol2[param::VoxZ]) {
+        const unsigned int colLim = (++row) + 1;
+        for (int iCol = 0; iCol < colLim; ++iCol) {
+          matrix(row, iCol) = cmatD[++iMat];
+          matrix(iCol, row) = matrix(row, iCol);
+        }
+      }
```

Comment:
Depending on the cpu needs, in AliRoot we had a fast access method to TMatrixT elements. @sawenzel , is there something similar in O2?

Reply:
In AliRoot Ruben had also written the AliSymMatrix class which was fast. I did not yet go deeper in checking the processing time for all the individual functions. I still need to check this and identify bottlenecks..

Reply:
I did not port AliSymMatrix to O2. Apart from the TMatrix, root provides the ``ROOT::Math::SMatrix`` and ``ROOT::Math::SVector`` classes which are supposed to be faster but are fixed dimension matrices (since the range of dimensions used in smoothing is limited, one can use just prebooked array of matrices of all needed dimensions). But I would propose to leave this optimizaton for the later stage.